cpp/src/arrow/python/inference.cc

```diff
@@ -354,12 +354,17 @@ class TypeInferrer {
       *keep_going = make_unions_;
     } else if (PyArray_CheckAnyScalarExact(obj)) {
       RETURN_NOT_OK(VisitDType(PyArray_DescrFromScalar(obj), keep_going));
-    } else if (PyList_Check(obj)) {
-      RETURN_NOT_OK(VisitList(obj, keep_going));
+    } else if (PySet_Check(obj) || (Py_TYPE(obj) == &PyDictValues_Type)) {
```

Comment:
Just wondering about a follow-up: could we optimize these checks somehow? 

Last time I ran the asv benchmarks the inference step had a significant overhead during the python to arrow conversions. 

@pitrou could we use a mapping or a different python API to eliminate these checks?

Reply:
It's true the list of consecutive checks started small and kept growing :-)
Using a mapping (keyed by `Py_TYPE(self)`, I suppose?) might be more efficient, but that remains to be verified. Also, you have to populate the mapping dynamically, to be sure you don't miss any cases. That is: use a mapping as a cache with an unbounded size, but fallback on this sequence of checks.

Reply:
Another possibility is to be opportunistic: if the current element was a `str`, then it's quite likely the next element will also be a `str`. So you can use a `switch/case` (keyed by an integer enum value?) to try to make the common case faster.

Reply:
Thanks Antoine! Created a follow-up to track this: https://issues.apache.org/jira/browse/ARROW-13914